Automatic script loading

chimicus opened the pull request:
A new script to traverse a full directory containing debug files for both the main binary and the shared objects. 

The script will:

for each object loaded in UDB check to see if a corresponding file is present in the directory specified by the user.
If there is such a file it will check for the Build-ID (if the UDB version comes with pyelftools) and if there is a match it will load the symbols.

barisione reviewed (changes requested):
Can you please format the file with `black` and add type annotations to the script?
You can use `./_linters/run-all.sh` to run the checks locally (this also happens on GitHub).

You attempt at dealing with a UDB without `elftools`, but we have another way of dealing with things in unreleased UDB versions.
There are two branches of this repo, `master` is supposed to work with the latest stable UDB release. `dev` is supposed to work with the next stable UDB release (that is, our internal `master`).

chimicus commented on automatic_symbol_loading/README.md:
```diff
@@ -0,0 +1,25 @@
+Load Debug Symbols
+==================
+
+A common practice amongst customers is to produce binaries compiled with debug symbols and
```
fixed

chimicus commented on automatic_symbol_loading/README.md:
```diff
@@ -0,0 +1,25 @@
+Usage: `load-all-symbols PATHTOBASEDIR`
+
+Examples:
+`load-all-symbols /data/mci/` - For each library loaded in UDB looks for the symbol file
```
yes, I thought it is generic enough so it can be used. Better not?

barisione replied:
It's OK but a bit unusual. What does mci stand for?
I less confusing path would be better, even something like `/path/to/symbols/root`.

chimicus commented on automatic_symbol_loading/README.md:
```diff
@@ -0,0 +1,25 @@
+
+Examples:
+`load-all-symbols /data/mci/` - For each library loaded in UDB looks for the symbol file
+(by taking into consideration all and only the files ending in '.debug') and, if the Build-IDs
```
fixed

chimicus commented on automatic_symbol_loading/README.md:
```diff
@@ -0,0 +1,25 @@
+(by taking into consideration all and only the files ending in '.debug') and, if the Build-IDs
+match, it loads the symbol-file.
+
+Note that the argument to `load-debug-symbols` needs to be a valid directory.
```
We do:
```
    def invoke(argument, from_tty):
        if not Path(argument).exists():
            print('Invalid directory specified')
            return
```

barisione replied:
Ah yes, I read the `README.md` first and forgot to update the comment.

barisione replied:
I think it's fine to remove this sentence. If it's not a valid directory I would not expect the user to expect it to work, and we do print a good error message.

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+import gdb
+from pathlib import Path
+
+check_build_id = True
```
done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+
+class ExtraSymbolsCommand(gdb.Command):
+    """
+    A command to load all external symbol files for debuggee
```
done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+    """
+
+    def __init__(self):
+        super().__init__("load-all-symbols", gdb.COMPLETE_EXPRESSION)
```
thanks, I wasn't aware

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+    @staticmethod
+    def invoke(argument, from_tty):
+        if not Path(argument).exists():
+            print('Invalid directory specified')
```
done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+                else:
+                    print(f'{leaf_name} has debug file: {sym_dict[leaf_name]} with mismatching Build-ID, ignoring')
+        if not check_build_id:
+            print('WARNING: couldn\' check the Build-ID of the loaded symbols. Results might be wrong. Update your copy of UDB to get pyelftools.')
```
both done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+        sym_dict = create_file_dict(argument)
+        for obj in gdb.objfiles():
+            leaf_name = Path(obj.filename).name
+            if leaf_name in sym_dict.keys():
```
good suggestion, done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+    p = sorted(Path(path).glob('**/*.debug'))
+    for f in p:
+        # Remove the .debug suffix from the file name
+        ret_dict[f.name[:-6]] = str(f)
```
done

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+        ret_dict[f.name[:-6]] = str(f)
+    return ret_dict
+
+def match_build_id(debug_file, obj):
```
done, note that one conversion is needed later on anyway: line 55:
```
                    obj.add_separate_debug_file(str(sym_dict[leaf_name]))
```
`add_separate_debug_file()` needs a string as input

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+        ef = ELFFile(fd)
+        sect = ef.get_section_by_name('.note.gnu.build-id')
+        # The data is a long string that contains the build-id at the end,
+        # starting from index 32
```
good point, added reference and a bit more explanation about 32. Ideally we could calculate "32" each time but I believe it is overshooting (see explanation).

chimicus commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,58 @@
+        debug_build_id = sect.data().hex()[32:]
+    if debug_build_id == obj.build_id:
+        return True
+    return False
```
fair enough

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,66 @@
+    """
+
+    def __init__(self):
+        super().__init__("load-all-symbols", gdb.COMPLETE_FILENAME)
```
It's still wrong. The completer argument is the third argument. The second indicates what kind of command this is.
Looking at https://sourceware.org/gdb/current/onlinedocs/gdb/Commands-In-Python.html, I think it should be a support command (`set sysroot` is a support command for instance) so:
```
super().__init__("load-all-symbols", gdb.COMMAND_SUPPORT, gdb.COMPLETE_FILENAME)
```

chimicus replied:
Apologies, I missed it. I think COMMAND_FILES might be better suited possibly as this is effectively mingling with files... what do you think?

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,66 @@
+    ret_dict = {}
+    p = sorted(Path(path).glob('**/*.debug'))
+    for f in p:
+        # Remove the .debug suffix from the file name
```
Missing full stop at the end of this comment.

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
```
No need for this in Python 3.

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
+from pathlib import Path
+import gdb
```
You should add a blank line above `import gdb` to separate standard and 3rd-party imports.

chimicus replied:
OK, added

barisione commented:
These were not addressed:

> Can you please format the file with `black` and add type annotations to the script?
> You can use `./_linters/run-all.sh` to run the checks locally (this also happens on GitHub).
> 
> You attempt at dealing with a UDB without `elftools`, but we have another way of dealing with things in unreleased UDB versions.
> There are two branches of this repo, `master` is supposed to work with the latest stable UDB release. `dev` is supposed to work with the next stable UDB release (that is, our internal `master`).

chimicus commented:
> These were not addressed:
> 
> > Can you please format the file with `black` and add type annotations to the script?
> > You can use `./_linters/run-all.sh` to run the checks locally (this also happens on GitHub).

Sorry, I did do a linting and I addressed the points that came up but I didn't add the type annotations. On it.


> > You attempt at dealing with a UDB without `elftools`, but we have another way of dealing with things in unreleased UDB versions.
> > There are two branches of this repo, `master` is supposed to work with the latest stable UDB release. `dev` is supposed to work with the next stable UDB release (that is, our internal `master`).

It is not clear to me how you would like me to address this (sorry If I'm being thick).

barisione commented:
> > > You attempt at dealing with a UDB without `elftools`, but we have another way of dealing with things in unreleased UDB versions.
> > > There are two branches of this repo, `master` is supposed to work with the latest stable UDB release. `dev` is supposed to work with the next stable UDB release (that is, our internal `master`).
> 
> It is not clear to me how you would like me to address this (sorry If I'm being thick).

You could just make this PR against `dev` so `elftools` is always available.
But if you think this is helpful even without it's fine to keep the workaround for now. We can drop it once a UDB with `elftools` in it is shipped.

chimicus commented:
> > > > You attempt at dealing with a UDB without `elftools`, but we have another way of dealing with things in unreleased UDB versions.
> > > > There are two branches of this repo, `master` is supposed to work with the latest stable UDB release. `dev` is supposed to work with the next stable UDB release (that is, our internal `master`).
> > 
> > 
> > It is not clear to me how you would like me to address this (sorry If I'm being thick).
> 
> You could just make this PR against `dev` so `elftools` is always available.
> But if you think this is helpful even without it's fine to keep the workaround for now. We can drop it once a UDB with `elftools` in it is shipped.

Given how quickly our users update their version of UDB I think it will be wise to keep this check at least for 6 months _after_ we release UDB with elftools.

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+    check_build_id = True
+
+
+def create_file_dict(path: str) -> dict:
```
Did you revert this to be a string? I thought you modified it to be a `Path`

barisione replied:
You should indicate also the type in the dictionary (`Dict[str, Path]`).

chimicus replied:
This is the function that changes the str and makes it become a Path, as it accepts in input the string passed by the user.
Adding the type int for the dictionary.

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+
+def create_file_dict(path: str) -> dict:
+    ret_dict = {}
+    p = sorted(Path(path).glob("**/*.debug"))
```
Why is it sorted?

barisione replied:
`p` is not a very useful name. What does it stand for?
Or you could just inline it in the `for`.

chimicus replied:
Sorted() is a left over of when I was experimenting, as it adds nothing I can remove it.

I will also remove "p" and inline it in the for

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+    Usage: allsym PATH_TO_SYMBOL_ROOT
+    """
+
+    def __init__(self):
```
Add `-> None:` otherwise this method is not type checked.

chimicus replied:
Added

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+
+    @staticmethod
+    def invoke(argument: str, from_tty: bool) -> None:
+        if not Path(argument).exists():
```
Just create a path object and use it later as well:
```
path = Path(argument)
if not path.exists():
    ...
sym_dict = create_file_dict(path)
```

chimicus replied:
yup, that's what I've done as part of above comment.

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+            raise gdb.GdbError("Invalid directory specified.")
+        sym_dict = create_file_dict(argument)
+        for obj in gdb.objfiles():
+            assert isinstance(obj.filename, str)
```
Why? Is this because mypy tells you that `filename` could be `None`?
If yes, then you need to handle it as it could actually happen to users!

chimicus replied:
Managed

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+            if debug_path is not None:
+                if match_build_id(debug_path, obj):
+                    print(f"Loading separate debug info for {leaf_name} from {sym_dict[leaf_name]}")
+                    obj.add_separate_debug_file(str(sym_dict[leaf_name]))
```
```
obj.add_separate_debug_file(str(debug_path))
```

chimicus replied:
yup changed

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,72 @@
+                    obj.add_separate_debug_file(str(sym_dict[leaf_name]))
+                else:
+                    print(
+                        f"{leaf_name} has debug file: {sym_dict[leaf_name]} "
```
s/`sym_dict[leaf_name]`/`debug_path`/

chimicus replied:
changed

barisione commented on automatic_symbol_loading/automatic_symbol_loading.py:
```diff
@@ -0,0 +1,77 @@
+    check_build_id = True
+
+
+def create_file_dict(path: Path) -> dict:
```
Sorry, I meant that the return type needs types specified, not `ret_dict`. `ret_dict` needs it only if mypy complains because it cannot infer it.

Without annotation, you have a properly typed dictionary inside the function but, for callers, the dictionary is just a mapping from anything to anything (i.e. `dict` is like saying `Dict[Any, Any]`).

chimicus replied:
Fair enough, I think I didn't think hard enough before fixing it :)
